ongov/ckanext-validation
# Transfer develop branch from external testing repo

## What this PR accomplishes
First working draft for integrating ckanext-validation with the core CKAN dictionary form as used in ontario-theme. Specifically:
- hard-codes custom checks according to agreed data quality rules around header name syntax and length, and some data value rules (bare numbers, bullet lists)
- validates resources in two phases: 
    - Phase 1 occurs on file upload when xloader tries to push it to database. Checks performed include the hard-coded rules and any postgres violations around tabular structure (e.g. missing headers, shifted rows). (Note that postgres violation checks come for free with the process of xloader trying to push to database). In Phase 1, there is no data dictionary, therefore data types are not checked.
    - Phase 2 occurs when the Validate button on the data dictionary page is clicked. This is when data types are checked against the data dictionary inputs.

No modifications made to the UI of ckanext-validation itself, so validation message boxes and validation forms should all look the same.

## What needs review
- code logic
- functionality


## How to test
- refer to internal documentation on setting up this branch with custom edits to the frictionless package and the latest pull from xloader master branch
- upload csv files that contain various errors to test the checks in Phase 1 and 2. Check that the errors are caught and are displayed, and that validation passes once the errors are fixed.
- test four different scenarios: 
    Case 1: brand new resource on Phase 1 (i.e. when xloader tries to push to database)
    Case 2:  same resource after is has passed Phase 1; test Phase 2 against the data dictionary form
    Case 3: update an existing resource that does NOT have any data dictionary defined (i.e. replace csv file with a modified file but do not create any data dictionary form)
    Case 4: update an existing resource that DOES have a data dictionary defined
